tamr_unify_client/client.py

```diff
@@ -1,26 +1,31 @@
 from urllib.parse import urljoin
 
 import requests
-from requests import Response
+import requests.exceptions
 
 from tamr_unify_client.dataset.collection import DatasetCollection
 from tamr_unify_client.project.collection import ProjectCollection
 
-# monkey-patch Response.successful
 
-
-def successful(self):
+def successful(response: requests.Response) -> requests.Response:
     """Checks that this response did not encounter an HTTP error (i.e. status code indicates success: 2xx, 3xx).
 
-    :raises :class:`requests.exceptions.HTTPError`: If an HTTP error is encountered.
-    :return: The calling response (i.e. ``self``).
-    :rtype: :class:`requests.Response`
+    Returns:
+        requests.Response: The response being checked.
+
+    Raises:
+        requests.exceptions.HTTPError: If an HTTP error is encountered (i.e. status code is 4xx or 5xx).
     """
-    self.raise_for_status()
-    return self
+    try:
+        response.raise_for_status()
+        return response
+    except requests.exceptions.HTTPError as e:
+        msg = str(e) + f"\nResponse body:\n{response.text}"
+        raise requests.exceptions.HTTPError(msg, response=response)
```

Comment:
It is kind of infuriating how non-standardized the Python `Exception` interface is.
My understanding was that the first argument is a message, which is intended to be a short summary text, and that further information is to be contained in other parts of the exception. This seems not only to be undocumented, but not the case - many Python Exceptions don't have a message (e.g. `OSError`).
`HTTPError` is only slightly less poorly documented than Exceptions in general; of note here is that, if no request is provided, `HTTPError.request` is filled in from `HTTPError.response.request`
So.
1. This information is already available in `e.response.text`; this is just promoting it up to the message. The message may become quite long as a result (e.g. HTTP 404 that returns an HTML body)
2. If the HTTPError contains a stack trace or other information, it will be lost in this construction.

Reply:
@nbateshaus The original error is not lost since previous exceptions in the chain can be accessed via `e.__context__` or `e.__cause__`, but I agree that we might want to handle (1) better. But that puts the onus on the user to write a `try`/`catch` block for EVERY `successful()` call. Perhaps we can log the error in some way instead? Maybe just write a "tamr-client.err.log" file from `cwd` and stick it in there?

Reply:
Promoting `e.response.text` to `e.args[0]` is not a bad thing. I *wish* there were a standard place to put more information, but there doesn't seem to be; since this seems like crucial information, it seems right to put it there.

Reply:
I think all of this works fine if `e.args[0]` is modified in-place to append `e.response.text`, and e is re-thrown.

Reply:
The issue with storing the response body in the exception is that if the user did not already have a `try`/`catch` block, the response body will no be recoverable and they will have to rerun after modifying their code. Plus the ergonomics of surrounding EVERY tamr client call with a `try`/`catch` block are not great.

The issue with printing out the response body is that the response body is that it goes against the "short summary" purpose of the message and could be an obnoxious amount of info to print on the terminal.

A solution that avoids both of these issues is to modify this code so that the original, unaltered `HTTPError` is re-raised AFTER we log the response body to a `tamr-client.log` file.

@nbateshaus : thoughts on this approach^^^?

Reply:
caught up with @nbateshaus offline. will proceed by replacing code in this PR to log (not print) the response body. users can configure logging to be to be to stdout, a file, etc..